strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+            "...j,...jk,...k", (vals - vc), np.linalg.inv(C + covmat), (vals - vc)
+        )
+        reweights = np.exp(-0.5 * reweights_exp_arg) / (
+            np.sqrt(np.linalg.det(2 * np.pi * (C + covmat)))
+        )
+        self.weights *= reweights
+        self.weights /= np.sum(self.weights)
+
+        self.means = self.means[abs(self.weights) > 0]
+        self.covs = self.covs[abs(self.weights) > 0]
+        self.weights = self.weights[abs(self.weights) > 0]
+
+    def post_select_homodyne(self, n, val, eps=0.0002, phi=0):
+        r"""Simulates a homodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            n (int): mode to be measured
+            val (array): measurement value to post-select
+            eps (int): squeezing of the measurement state
+            phi (float): homodyne angle
+        """
+        if self.active[n] is None:
+            raise ValueError("Cannot apply homodyne measurement, mode does not exist")
+        self.phase_shift(phi, n)
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        indices = [n]
+        vals = np.array([val, 0])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def post_select_heterodyne(self, n, alpha_val):
+        r"""Simulates a heterodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            n (int): mode to be measured
+            alpha_val (array): measurement value to post-select
+        """
+        if self.active[n] is None:
+            raise ValueError("Cannot apply heterodyne measurement, mode does not exist")
+
+        covmat = self.hbar * np.identity(2) / 2
+        indices = [n]
+        vals = np.array([alpha_val.real, alpha_val.imag])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def apply_u(self, U):
+        r"""Transforms the state according to the linear optical unitary that
+        maps a[i] \to U[i, j]^*a[j].
+
+        Args:
+            U (array): linear opical unitary matrix
+        """
+        Us = symp.interferometer(U)
+        self.means = update_means(self.means, Us, self.from_xp)
+        self.covs = update_covs(self.covs, Us, self.from_xp)
+
+    def apply_channel(self, X, Y):
+        r"""Transforms the state according to a deterministic Gaussian CPTP map.
+
+        Args:
+            X (array): matrix for mutltiplicative part of transformation
+            Y (array): matrix for additive part of transformation.
+        """
+        self.means = update_means(self.means, X, self.from_xp)
+        self.covs = update_covs(self.covs, X, self.from_xp, Y)
+
+    def expandS(self, modes, S):
+        """Expands symplectic matrix on subset of modes to symplectic matrix for the whole system.
+
+        Args:
+            modes (list): list of modes on which S acts
+            S (array): symplectic matrix
+        """
+        return symp.expand(S, modes, self.nlen)
+
+    def expandXY(self, modes, X, Y):
```

Comment:
can't you do this by calling `symp.expand` twice?

Reply:
I think I get the point now, for `Y2` you pad with zeros even along the diagonal.

Reply:
You could do that explicitly though, by calling `symp.expand` and then doing a for loop loop over the diagonal elements of the untouched modes.

Reply:
I'd still prefer to use `expand` and set the diagonal elements to zero than do all of this numpy indexing black magiv.

Reply:
Might also be worth adding a specific test case for this

Reply:
In the test for loss, we test creating a thermal state via the thermal loss method which applied this method directly.

Reply:
The `thermal_loss` method does some other things too (initiates a particular `X` and `Y`, calls this method and calls `apply_chanel`). Writing a unit test for that can have a different effect to writing a unit test just for `expandXY`.

Reply:
Done.